### ISSUE TITLE
Feature Request: Avoid panicking when ring and aws_lc_rs are both specified

### ISSUE THREAD

**Issue by Alvenix** [rustls/rustls#1877]
**Checklist**
* [x] I've searched the issue tracker for similar requests

I know the current design for crypto providers is probably well discussed but I wanted to add one data point against the current behavior for reference. This issue is mostly rephrasing my discussion at [1876](https://github.com/rustls/rustls/issues/1876#issuecomment-2023403756):

**Is your feature request related to a problem? Please describe.**
When I accidentally enabled these two features (directly or indirectly through a crate) I had few panics that I had to debug. I consider myself somewhat familiar with Rust and (and cargo features) so debugging this was easy although annoying. However, this could be a little road blocker for people less familiar with Rust.

**Describe the solution you'd like**
Changing the internal function from_crate_features() to choose either `ring` or `aws_lc_rs` CryptoProvider when both features are specified, instead of returning None. In addition, removing the related panic. In addition, specify the behavior so it is no longer ambiguous.

**Describe alternatives you've considered**
Setup the default crypto provider on main. However, sometimes when developing a library I probably shouldn't do that. I would have to document to the users, that they need to setup the crypto providers if they don't want to have a panic accidentally.

**Comment by djc** [comment]
I agree that the current setup is not ideal for libraries. I think we should have gone with something like this:

```diff
diff --git a/rustls/src/crypto/mod.rs b/rustls/src/crypto/mod.rs
index 74b19ea6..9f029aed 100644
--- a/rustls/src/crypto/mod.rs
+++ b/rustls/src/crypto/mod.rs
@@ -215,6 +215,29 @@ pub struct CryptoProvider {
 }
 
 impl CryptoProvider {
+    /// Returns the default `CryptoProvider` for this process.
+    ///
+    /// If a crypto provider has been installed using [`install_default()`], return that.
+    /// Otherwise, return the default provider for the crate features in use. If both
+    /// first-party providers are enabled, this will return the aws-lc-rs provider.
+    #[cfg(any(feature = "aws_lc_rs", feature = "ring"))]
+    pub fn installed_or_default() -> Arc<Self> {
+        match PROCESS_DEFAULT_PROVIDER.get() {
+            Some(provider) => provider.clone(),
+            None => {
+                #[cfg(feature = "aws_lc_rs")]
+                {
+                    return Arc::new(aws_lc_rs::default_provider());
+                }
+
+                #[cfg(all(feature = "ring", not(feature = "aws_lc_rs")))]
+                {
+                    return Arc::new(ring::default_provider());
+                }
+            }
+        }
+    }
+
     /// Sets this `CryptoProvider` as the default for this process.
     ///
     /// This can be called successfully at most once in any process execution.
```

And used this in the builders. This would never panic, because it is only available when one of the first-party providers is enabled, but it would require restricting the builder constructors to only be available when at least one of the first-party providers is enabled.

For now, we could still provider this as a helper for libraries. We could also make a change to avoid panicking in the case where both first-party providers are available, and pick aws-lc-rs in that case, which seems reasonable to me. The current setup breaks feature additivity which is a bit of a pain.

I also wonder whether it is the right choice to implicitly install a crypto provider in `get_default_or_install_from_crate_features()` -- instead, it could just return a provider if no default has been installed.

**Comment by jbr** [comment]
Not sure if this is the right place for this feedback, but I accidentally enabled both features while switching over some code from previous default (ring) to new default (aws-ls-rs) and almost didn't catch this because I was running cargo check instead of executing the code. If having both features enabled is a show-stopping error (panic in likely-unrecoverable code) and the information is available at compile time (cargo features), it would be more ergonomic to emit a `compile_error` instead of panicking at runtime in `get_default_or_install_from_crate_features`.

**Comment by ctz** [comment]
> The current setup breaks feature additivity which is a bit of a pain.

That proposal also breaks feature additivity: if you carefully arrange to enable only the `ring` feature because you want to use ring, but later something in your distant dependency enables `aws-lc-rs` too, you silently end up using  `aws-lc-rs` against your wishes.

The purpose of the panic is to signal clearly: "you have failed to unambiguously select a provider via crate features."

I think my first preference is to see if any progress is made with the ["mutually-exclusive, global features" proposal](https://internals.rust-lang.org/t/pre-rfc-mutually-excusive-global-features/19618).

If not, I think we could look into making the default provider thing fail at link time, perhaps. That would look something like:

(in the rustls crate)
```
#[cfg(feature = "aws-lc-rs")]
specify_default_provider! ( &aws_lc_rs::PROVIDER );

#[cfg(feature = "ring")]
specify_default_provider! ( &ring::PROVIDER );

extern "C" {
  pub static RUSTLS_DEFAULT_PROVIDER: &'static CryptoProvider;
}
```

The `specify_default_provider!()` macro would expand to emitting a definition of `extern "C" RUSTLS_DEFAULT_PROVIDER`. The final link would fail if:

- anything refers to `RUSTLS_DEFAULT_PROVIDER`, but nothing provides it (ie, you have used `ClientConfig::builder()`) -- via a missing symbol error
- multiple `specify_default_provider!` are in play, with a multiple definition error

But linker errors are exceedingly user-unfriendly, and reading `RUSTLS_DEFAULT_PROVIDER` would introduce `unsafe` code into the crate.

**Comment by NobodyXu** [comment]
Would it be more reasonable to have `compile_error!`, instead of a runtime panic when both are specified?

**Comment by djc** [comment]
The caller can bypass this code path by using [`builder_with_provider()`](https://docs.rs/rustls/latest/rustls/client/struct.ClientConfig.html#method.builder_with_provider) instead.

**Comment by cpu** [comment]
I don't think this issue surfaced a workable alternative we could implement that doesn't have its own new set of downsides. I'm personally leaning towards saying we should close the issue for now. WDYT?

**Comment by cpu** [comment]
I think that proposal is unworkable based on [Ctz's point](https://github.com/rustls/rustls/issues/1877#issuecomment-2025981904):

> That proposal also breaks feature additivity: if you carefully arrange to enable only the ring feature because you want to use ring, but later something in your distant dependency enables aws-lc-rs too, you silently end up using aws-lc-rs against your wishes.

**Comment by divergentdave** [comment]
I think it's important to address this issue, as it is a barrier to ecosystem adoption in other libraries. 

What if rustls had separate Cargo features to enable each provider and select each provider as the default? The documentation could tell end users to enable one default selection feature, and tell libraries to only use the features that enable providers. This would still leave one case where initialization could panic (if neither or both default selection features were enabled, and there wasn't a runtime default provider installed) but this could now be solved either by using `install_default()` or changing the end user-only feature flags. (In particular, adding a dev-dependency that picks a default provider feature would be more ergonomic for unit tests than calling `install_default()` in each.) Separating the dual purposes of the feature flags this way would mean the flags that enable providers would be truly additive. If rustls still included one of the "default provider selection" features in its `default` feature, we would still have the problem we have today of a few library crates forgetting to set no-default-features and causing problems. If not, then there would no longer be a default choice of a default, and end users would need to read the documentation and add a feature as an additional barrier to entry, but there would be better adherence by library crates to the guidance to not select a default provider.

**Comment by djc** [comment]
Personally I find @Alvenix's line of reasoning pretty compelling. As a downstream library maintainer I haven't wanted to adopt the installed default if that means I can incur a panic, and if downstream maintainers don't adopt it it's pretty pointless to have the option at all.

**Comment by rbtcollins** [comment]
What support is needed to move this forward?

**Comment by djc** [comment]
> What support is needed to move this forward? 

In the limit, Cargo support for global and/or mutually exclusive features.

**Comment by neoeinstein** [comment]
Dropping in a note here that the current behavior is quite annoying. In our case, it requires adding in a dependency on the specific 0.23.x line to be able to install one of the crypto providers as default. But, this will all likely break again on 0.24.x, which means that I'll need to install _another_ shim for 0.24.x. I won't have an easy way of knowing when those shims can be removed either.

My preference would be for rustls to just _have_ a default if both are selected, and use application-oriented feature flags to set the default or force one if that ambiguity actually matters: `default_provider_ring` and `default_provider_aws_lc_rs` or `force_provider_ring` and `force_provider_aws_lc_rs`.

Many library authors end up selecting either `ring` or `aws_lc_rs` to ensure that _some_ provider is configured, precisely to avoid that unnecessarily rising to an application level concern. The problem comes when those are in conflict. Perhaps adding a feature flag that would allow library authors to indicate: "I don't really care about the underlying crypto library, I just need _something_", instead of only `ring` and `aws_lc_rs`. That would then allow libraries to pull in `rustls` without needing to make an explicit selection if they don't have skin in that game, while still allowing an application-level consumer to make an explicit selection to override a default if that matters to them.

My current situation is lodged between the `aws-smithy-http-client` having newly enabled `aws_lc_rs` and `tonic` having long enabled `ring`, and I have no positive care which is used, but if I start depending directly on `rustls` to resolve the disambiguation, I'm going to have to do that individually for each breaking version that `rustls` releases as different parts update at different rates.

**Comment by estk** [comment]
@djc This panic happened to us in production. We certainly dropped the ball by only having http tests, but in our situation this really does seem like a bug.

## Context:

- Write custom tls verifier with ring/rustls before aws-lc-rs exists because our customer cant get it together
- Test and deploy working
- Update this crate, all good, we just default-features = false and features = "ring"
- Test and deploy working
- 3rd party crate update, it updates rustls version and omits no-default-features
- Broken at runtime

The thing that is really strange is that rustls has effectively selected a default already **by including aws-lc-rs in default-features, in a crate with a large dep graph you just cant rely on the absence of features**, therefore panicing on the addition of a feature is a bug in my mind.

## Potential fixes

- Add a new feature: default-verifier-rusts, if its set, thats the default verifier you get, if not, aws-lc-rs
- Add a custom cfg flag
  -  https://docs.rs/loom/latest/loom/#running-loom-tests
- Add an env var checked at compile time ex RUSTLS_DEFAULT_VERIFIER=ring
- Make the only way to initialize CryptoProviders thru a macro that appends to a custom linker section that at least would be statically verifiable.
  - https://github.com/mmastrac/rust-ctor
  - https://github.com/dtolnay/inventory

**Comment by dt-spr** [comment]
This misuse of features completely breaks the "if it compiles, it works" mantra of Rust.

# Potential fix: just be opinionated

The use of `ClientConfig::builder()` indicates that the client is not opinionated about the default provider. You cannot expect all clients to know what's the best provider, this is why defaults matter.

Clearly `aws-lc-rs` has been selected as a default after careful deliberation. And in the future, `rustls` may pick another provider if `aws-lc-rs` ends up unmaintained, or has unfixed vulnerabilities, etc... Users are trusting _you_, rustls maintainers, to keep abreast of the state of the art and to pick the right default.

So just make it work.

In `CryptoProvider::from_crate_features` test the features from "best" to "worst" according to _your_ ranking, and pick the first that works:

```rust
impl CryptoProvider {
    fn from_crate_features() -> Option<Self> {
        #[cfg(feature = "aws_lc_rs")]
        {
            return Some(aws_lc_rs::default_provider());
        }

        #[cfg(feature = "ring")]
        {
            return Some(ring::default_provider());
        }

        #[allow(unreachable_code)]
        None
    }
}
```

**Advantages**:

- Features are additive again.
- Defaults just work.

**Disadvantages**:

- A client may expect `ring`, or a custom provider, which used to be the default, and not get it.
- Implicit, like all defaults tend to be.

**Comment by djc** [comment]
We're not going to change this for 0.23.x. However, we're currently discussing how we want to change this API in 0.24. Feel free to participate in that discussion here:

- #2396
